python/podio/pythonizations/collection_subscript.py

```diff
@@ -0,0 +1,27 @@
+"""Pythonize subscript operation for collections"""
+
+import cppyy
+from .utils.pythonizer import Pythonizer
+
+
+class CollectionSubscriptPythonizer(Pythonizer):
+    """Bound-check __getitem__ for classes derived from podio::CollectionBase"""
+
+    @classmethod
+    def priority(cls):
+        return 50
+
+    @classmethod
+    def callback(cls, class_, name):
```

Comment:
So every callback runs once per object in the namespace they are tied to? That would mean every EDM4hep object, collection or not

Reply:
Once registered, cppyy will try to lazily apply the callback to all the entities in a given namespace. So the callbacks usually start with a conditional to filter out when they should be applied.
Yes, it creates a small overhead for importing any definition from that namespace

Reply:
It's being called more often than one would think, for example, after adding a print in the callback:
```
In [1]: import edm4hep

In [2]: edm4hep.<TAB>
Calling callback from collection_subscript.py

In [2]: coll = edm4hep.MCParticleCollection()
Calling callback from collection_subscript.py

In [3]: coll.create()
Calling callback from collection_subscript.py
Out[3]: <cppyy.gbl.edm4hep.MutableMCParticle object at 0x59e0115f8cd0>

In [4]: edm4hep.<TAB>
Calling callback from collection_subscript.py
Calling callback from collection_subscript.py
Calling callback from collection_subscript.py
Calling callback from collection_subscript.py
```
and after creating more objects `edm4hep.<TAB>` will call the callback many many times. I'm not sure if much can be done about that, otherwise one has to be careful what is put in the callbacks or some parts of using python interactively will become unusable.

Reply:
Thanks for checking. I get a slightly different behavior and the callback seems to be indeed called only once for each class in namespace and only when cppyy loads it for the first time
```python
>>> import edm4hep
>>> edm4hep.<TAB>
Calling callback for Quantity
Calling callback for RawTimeSeriesData
Calling callback for MutableMCRecoCaloAssociation
...
Display all 123 possibilities? (y or n)
edm4hep.CaloHitContribution(
edm4hep.CaloHitContributionCollection(
edm4hep.CaloHitContributionData(
...
>>
>>> coll = edm4hep.MCParticleCollection()
>>> coll.create()
<cppyy.gbl.edm4hep.MutableMCParticle object at 0x99d8b70>
>>> edm4hep.<TAB>
Display all 123 possibilities? (y or n)
edm4hep.CaloHitContribution(
edm4hep.CaloHitContributionCollection(
edm4hep.CaloHitContributionData(
```
with modified callback
```python
    @classmethod
    def callback(cls, class_, name):
        def get_item(self, i):
            try:
                return self.at(i)
            except cppyy.gbl.std.out_of_range:
                raise IndexError("collection index out of range") from None

        print(f"Calling callback for {name}")
        if issubclass(class_, cppyy.gbl.podio.CollectionBase) and hasattr(class_, "at"):
            class_.__getitem__ = get_item
```

Reply:
Could this be due to some `%autoreload` magic? I could imagine that would influence how often the callback is called.

Reply:
I see. So the difference is that plain python's tab-completion on a module forces cppyy to load everything now but ipython's loads only a few entities now and the rest is loaded lazily. In both scenarios each callback will be called at most once for each class. I was worried that somehow the callbacks are called over and over again for the same things, but it's not the case here.

It seems that `%autoreload`  doesn't reset the state of cppyy. So neither the shared libraries are read again nor callbacks applied again